### PR TITLE
Send mqtt message when motion is detected

### PR DESCRIPTION
**Pull request by NickM-27**
Adds a motion topic that can be used as a generic motion detector.

### CODE REVIEW

**Comment by blakeblackshear** [comment]
You can't pass the mqtt client to a separate process like this. It would need to be handled in object_processing.py along with the other mqtt sensors.

**Comment by NickM-27** [comment]
> You can't pass the mqtt client to a separate process like this. It would need to be handled in object_processing.py along with the other mqtt sensors.

Awesome, ended up being a lot cleaner as well!

**Comment by NickM-27** [comment]
Not sure why mypy is all of the sudden failing on plus.py ?

**Comment by hawkeye217** [comment]
What do you think about adding an optional config option under the ``mqtt:`` section for each camera to enable or disable this? It'd be nice to be able to mitigate excess mqtt traffic in case someone doesn't need the feature.

**Comment by NickM-27** [comment]
> What do you think about adding an optional config option under the `mqtt:` section for each camera to enable or disable this? It'd be nice to be able to mitigate excess mqtt traffic in case someone doesn't need the feature.

The mqtt section is difficult since currently that is directly related to images. Might need to add a `messages` section under there to configure these and other messages in the future.

**Comment by hawkeye217** [comment]
> I already have the code setup such that it shouldn't send redundant messages with the same state, so I am not sure how much extra traffic this will add especially with the default 30 seconds off timeout.

Right. I think it should be good then.

**Comment by NickM-27** [comment]
Just as a point of reference, here is MQTT explorer showing the sensor working as expected:

Screen Shot 2022-05-01 at 11 19 58 AM

**Comment by blakeblackshear** [comment]
Don't the other sensors publish ON and OFF?

**Comment by NickM-27** [comment]
> Don't the other sensors publish ON and OFF?

That's for the switches / toggles though, I don't believe we have any other sensors that publish a state so I was thinking "motion detected is ON" wasn't intuitive, but I don't feel strongly so if the consistency is better, then`ON` works as well

**Comment by NickM-27** [comment]
@blakeblackshear after thinking about it a bit more I can think of 3 good options:

1. Keep it as it is now: `/motion/detected` = `True`/`False`
2. Change it to `/motion/detected` = `ON`/`OFF`
3. Change it to `/motion/status` = `DETECTED`/`CLEAR`

let me know what you think 👍

**Comment by blakeblackshear** [comment]
If it will translate to an mqtt binary sensor in home assistant, I would prefer to stick the default expected payloads. That's why ON/OFF have been used historically. https://www.home-assistant.io/integrations/binary_sensor.mqtt/

File: docs/docs/configuration/index.md [hunk 1]
```diff
@@ -246,6 +246,8 @@ motion:
   # Enables dynamic contrast improvement. This should help improve night detections at the cost of making motion detection more sensitive
   # for daytime.
   improve_contrast: False
+  # Optional: Delay when updating camaera motion through MQTT from True -> False. (default: shown below)
```
Comment (blakeblackshear) [inline docs/docs/configuration/index.md:249]: Typo. `camaera`. Also the comment should show `ON` and `OFF` now.

File: frigate/object_processing.py [hunk 2]
```diff
@@ -844,6 +845,33 @@ def should_mqtt_snapshot(self, camera, obj: TrackedObject):
...
+            # always updated latest motion
+            self.last_motion_updates[camera] = int(time.time())
+        elif not motion_boxes and self.last_motion_updates.get(camera, 0) != 0:
+            mqtt_delay = self.config.cameras[camera].motion.mqtt_off_delay
+            now = int(time.time())
+
+            # If no motion, make sure the off_delay has passed
+            if now - self.last_motion_updates.get(camera, 0) >= mqtt_delay:
+                self.client.publish(
+                    f"{self.topic_prefix}/{camera}/motion/detected",
```
Comment (blakeblackshear) [inline frigate/object_processing.py:868]: Let's make the topic `f"{self.topic_prefix}/{camera}/motion"` instead to be more consistent with other topic names.

File: frigate/object_processing.py [hunk 3]
```diff
@@ -844,6 +845,33 @@ def should_mqtt_snapshot(self, camera, obj: TrackedObject):
...
+                    "ON",
+                    retain=False,
+                )
+
+            # always updated latest motion
+            self.last_motion_updates[camera] = int(time.time())
+        elif not motion_boxes and self.last_motion_updates.get(camera, 0) != 0:
+            mqtt_delay = self.config.cameras[camera].motion.mqtt_off_delay
+            now = int(time.time())
```
Comment (blakeblackshear) [inline frigate/object_processing.py:863]: In the rest of the codebase, I use `datetime.datetime.now().timestamp()` to get the current time. Let's use that here for consistency.

**Review by blakeblackshear — CHANGES_REQUESTED** [review]

File: docs/docs/integrations/mqtt.md [hunk 4]
```diff
@@ -123,6 +123,11 @@ Topic with current state of snapshots for a camera. Published values are `ON` an
 Topic to turn motion detection for a camera on and off. Expected values are `ON` and `OFF`.
 NOTE: Turning off motion detection will fail if detection is not disabled.
 
+### `frigate/<camera_name>/motion`
+
+Whether camera_name is currently detecting motion. Expected valeus are `ON` and `OFF`.
+NOTE: Will reset to `OFF` after configurable amount of time (30 seconds by default).
```
Comment (blakeblackshear) [inline docs/docs/integrations/mqtt.md:129]: This makes it sound like it resets to off after 30 seconds always. Can you clarify that it resets to off 30 seconds after the last time motion was detected?

File: docs/docs/integrations/mqtt.md [hunk 5]
```diff
@@ -123,6 +123,12 @@ Topic with current state of snapshots for a camera. Published values are `ON` an
 Topic to turn motion detection for a camera on and off. Expected values are `ON` and `OFF`.
 NOTE: Turning off motion detection will fail if detection is not disabled.
 
+### `frigate/<camera_name>/motion`
+
+Whether camera_name is currently detecting motion. Expected values are `ON` and `OFF`.
+NOTE: After motion is initially detected, `ON` will be set until no motion has
+been detected for `motion_off_delay` seconds (30 by default).
```
Comment (blakeblackshear) [inline docs/docs/integrations/mqtt.md:130]: The setting is mqtt_off_delay, not motion_off_delay.

**Review by blakeblackshear — APPROVED** [review]

File: docs/docs/integrations/mqtt.md (hunk 5 above)
Comment (NickM-27) [reply docs/docs/integrations/mqtt.md]: Oh dang, silly mistake, thanks for fixing

**Pull request merged**